File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        charSet = set()
        
        def overlap(cSet,word):
            prev = set()
            for c in word:
                if c in cSet or c in prev:
                    return True
                prev.add(c)
            return False
        
        def backtrack(i):
            if i == len(arr):
                return len(charSet)
            res = 0
            if not overlap(charSet,arr[i]):
                for c in arr[i]:
                    charSet.add(c)
                res = backtrack(i+1)
                for c in arr[i]:
                    charSet.remove(c)
            return max(res,backtrack(i+1))
        
        return backtrack(0)

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.cpp
class Solution {
public:
    unordered_set<char> charSet;
    bool overlap(unordered_set<char> cSet, string word) {
        unordered_set<char> prev;
        for(char c: word) {
            if(charSet.find(c) != charSet.end() or prev.find(c) != prev.end()) {
                return true;
            }
            prev.insert(c);
        }
        return false;
    }
    
    int backtrack(int i,vector<string>& arr) {
        if (i == arr.size()) {
            return charSet.size();
        }
        int res = 0;
        if (!overlap(charSet,arr[i])) {
            for(char c: arr[i]) {
                charSet.insert(c);
            }
            res = backtrack(i+1,arr);
            for(char c: arr[i]) {
                charSet.erase(c);
            }
        }
        return max(res,backtrack(i+1,arr));
    }
    
    int maxLength(vector<string>& arr) {
        return backtrack(0,arr);    
    }
};